api/rest/bundle_handler.go

```diff
@@ -0,0 +1,354 @@
+			bundle.Stopped = h.clock.Now()
+			newBundleStatus := jsonMarshal(bundle)
+			e := ioutil.WriteFile(stateFilePath, newBundleStatus, filePerm)
+			if e != nil {
+				logrus.WithError(e).Errorf("Could not update state file %s", id)
+			}
+		}
+	}()
+
+	write(w, bundleStatus)
+}
+
+func collectAll(ctx context.Context, done chan<- []string, dataFile io.WriteCloser, collectors []collectors.Collector) {
+	zipWriter := zip.NewWriter(dataFile)
+	var errors []string
+	// summaryReport is a log of a diagnostics job
+	summaryReport := new(bytes.Buffer)
+
+	for _, c := range collectors {
+		if ctx.Err() != nil {
+			errors = append(errors, ctx.Err().Error())
+			break
+		}
+		summaryReport.WriteString(fmt.Sprintf("[START GET %s]\n", c.Name()))
+		err := collect(ctx, c, zipWriter)
+		summaryReport.WriteString(fmt.Sprintf("[STOP GET %s]\n", c.Name()))
+		if err != nil && !c.Optional() {
+			errors = append(errors, err.Error())
+		}
+	}
+
+	summaryReportFile, err := zipWriter.Create(summaryReportFileName)
+	if err != nil {
+		errors = append(errors, err.Error())
+	} else {
+		if _, err := io.Copy(summaryReportFile, summaryReport); err != nil {
+			errors = append(errors, err.Error())
+		}
+	}
+
+	if len(errors) != 0 {
+		summaryErrorReportFile, err := zipWriter.Create(summaryErrorsReportFileName)
+		if err != nil {
+			errors = append(errors, err.Error())
+		} else {
+			if _, err := summaryErrorReportFile.Write([]byte(strings.Join(errors, "\n"))); err != nil {
+				errors = append(errors, err.Error())
+			}
+		}
+	}
+
+	if err := zipWriter.Close(); err != nil {
+		errors = append(errors, err.Error())
+	}
+	if err := dataFile.Close(); err != nil {
+		errors = append(errors, err.Error())
+	}
+
+	done <- errors
+}
+
+func collect(ctx context.Context, c collectors.Collector, zipWriter *zip.Writer) error {
+	rc, err := c.Collect(ctx)
+	if err != nil {
+		return fmt.Errorf("could not collect %s: %s", c.Name(), err)
+	}
+	defer rc.Close()
+
+	zipFile, err := zipWriter.Create(c.Name())
+	if err != nil {
+		return fmt.Errorf("could not create a %s in the zip: %s", c.Name(), err)
+	}
+	if _, err := io.Copy(zipFile, rc); err != nil {
+		return fmt.Errorf("could not copy data to zip: %s", err)
+	}
+
+	return nil
+}
+
+func (h BundleHandler) Get(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["uuid"]
+
+	bundle, err := h.getBundleState(id)
+	if err != nil {
+		writeJSONError(w, http.StatusNotFound, fmt.Errorf("bundle not found: %s", err))
+		return
+	}
+
+	body := jsonMarshal(bundle)
+	write(w, body)
+}
+
+func (h BundleHandler) GetFile(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["uuid"]
+
+	http.ServeFile(w, r, filepath.Join(h.workDir, id, dataFileName))
+}
+
+func (h BundleHandler) List(w http.ResponseWriter, r *http.Request) {
+	ids, err := ioutil.ReadDir(h.workDir)
+	if err != nil {
+		writeJSONError(w, http.StatusInsufficientStorage, fmt.Errorf("could not read work dir: %s", err))
```

Comment:
same as above, 500 is the better choice since we don't know the cause for this error.